File: ts/src/color-conv.ts
/**
 * Color Conversion Utilities for OKLCH ↔ sRGB.
 * Reference: https://bottosson.github.io/posts/oklab/
 */

import type { Rgba, OklchComponents } from './types.js';

/**
 * Converts OKLCH values to sRGB.
 * @param l - Lightness (0-1)
 * @param c - Chroma (0-~0.4)
 * @param h - Hue (0-360)
 * @param a - Alpha (0-1), default 1
 * @returns {r, g, b, a} in 0-255 range
 */
export function oklchToRgb(l: number, c: number, h: number, a: number = 1): Rgba {
    const hr = (h * Math.PI) / 180;
    const aCoord = c * Math.cos(hr);
    const bCoord = c * Math.sin(hr);

    // OKLab → LMS (cone response)
    const l_ = l + 0.3963377774 * aCoord + 0.2158037573 * bCoord;
    const m_ = l - 0.1055613458 * aCoord - 0.0638541728 * bCoord;
    const s_ = l - 0.0894841775 * aCoord - 1.291485548 * bCoord;

    // Cube LMS values (clamped to prevent negative)
    const L = Math.pow(Math.max(0, l_), 3);
    const M = Math.pow(Math.max(0, m_), 3);
    const S = Math.pow(Math.max(0, s_), 3);

    // LMS³ → Linear sRGB
    let r = +4.0767416621 * L - 3.3077115913 * M + 0.2309699292 * S;
    let g = -1.2684380046 * L + 2.6097574011 * M - 0.3413193965 * S;
    let b = -0.0041960863 * L - 0.7034186147 * M + 1.707614701 * S;

    // sRGB gamma correction
    const gamma = (x: number): number =>
        x <= 0.0031308 ? 12.92 * x : 1.055 * Math.pow(x, 1 / 2.4) - 0.055;

    // Clamp and scale to 0-255
    const rOut = Math.round(Math.min(1, Math.max(0, gamma(r))) * 255);
    const gOut = Math.round(Math.min(1, Math.max(0, gamma(g))) * 255);
    const bOut = Math.round(Math.min(1, Math.max(0, gamma(b))) * 255);
    const aOut = Math.round(Math.min(1, Math.max(0, a)) * 255);

    return { r: rOut, g: gOut, b: bOut, a: aOut };
}

/**
 * Formats RGB to 6-digit Hex (e.g. #FFFFFF).
 */
export function toHex6(rgba: { r: number; g: number; b: number }): string {
    const r = rgba.r.toString(16).padStart(2, '0').toUpperCase();
    const g = rgba.g.toString(16).padStart(2, '0').toUpperCase();
    const b = rgba.b.toString(16).padStart(2, '0').toUpperCase();
    return `#${r}${g}${b}`;
}

/**
 * Formats RGB to 8-digit Hex with Alpha (e.g. 0XFFFFFFFF).
 * Used by Flutter.
 */
export function toHex8(rgba: Rgba): string {
    const alpha = rgba.a.toString(16).padStart(2, '0').toUpperCase();
    const r = rgba.r.toString(16).padStart(2, '0').toUpperCase();
    const g = rgba.g.toString(16).padStart(2, '0').toUpperCase();
    const b = rgba.b.toString(16).padStart(2, '0').toUpperCase();
    return `0X${alpha}${r}${g}${b}`;
}

/**
 * Parses an oklch(...) string into its components and RGB values.
 * @returns Parsed components or null if the string is invalid.
 */
export function parseOklch(str: string): OklchComponents | null {
    const match = str.match(
        /oklch\(\s*([\d.]+)%?\s+([\d.]+)\s+([\d.]+)(?:\s*\/\s*([\d.]+))?\s*\)/i
    );
    if (!match) return null;

    const l = parseFloat(match[1]!) / 100;
    const c = parseFloat(match[2]!);
    const h = parseFloat(match[3]!);
    const a = match[4] ? parseFloat(match[4]) : 1;

    const rgb = oklchToRgb(l, c, h, a);
    return {
        ...rgb,
        l,
        c,
        h,
        opacity: a,
    };
}

File: dart/lib/src/color_conv.dart
/// OKLCH to RGB conversion and utilities.
import 'dart:math' as math;
import 'types.dart';

RgbaColor oklchToRgb(double l, double c, double h, [double a = 1.0]) {
  final hr = h * math.pi / 180.0;
  final aCoord = c * math.cos(hr);
  final bCoord = c * math.sin(hr);

  final l_ = l + 0.3963377774 * aCoord + 0.2158037573 * bCoord;
  final m_ = l - 0.1055613458 * aCoord - 0.0638541728 * bCoord;
  final s_ = l - 0.0894841775 * aCoord - 1.2914855480 * bCoord;

  // Signed cubing
  final l3 = l_ * l_ * l_;
  final m3 = m_ * m_ * m_;
  final s3 = s_ * s_ * s_;

  final rLin = 4.0767416621 * l3 - 3.3077115913 * m3 + 0.2309699292 * s3;
  final gLin = -1.2684380046 * l3 + 2.6097574011 * m3 - 0.3413193965 * s3;
  final bLin = -0.0041960863 * l3 - 0.7034186147 * m3 + 1.7076147010 * s3;

  double toSrgb(double x) {
    if (x <= 0.0) return 0.0;
    if (x >= 1.0) return 1.0;
    return x <= 0.0031308 ? 12.92 * x : 1.055 * math.pow(x, 1 / 2.4) - 0.055;
  }

  return RgbaColor(
    r: (toSrgb(rLin) * 255.0).round().clamp(0, 255),
    g: (toSrgb(gLin) * 255.0).round().clamp(0, 255),
    b: (toSrgb(bLin) * 255.0).round().clamp(0, 255),
    a: a,
  );
}

String toHex6(RgbaColor rgba) {
  final r = rgba.r.toRadixString(16).padLeft(2, '0');
  final g = rgba.g.toRadixString(16).padLeft(2, '0');
  final b = rgba.b.toRadixString(16).padLeft(2, '0');
  return '#${(r + g + b).toUpperCase()}';
}

String toHex8(RgbaColor rgba) {
  final a =
      (rgba.a * 255.0).round().clamp(0, 255).toRadixString(16).padLeft(2, '0');
  final r = rgba.r.toRadixString(16).padLeft(2, '0');
  final g = rgba.g.toRadixString(16).padLeft(2, '0');
  final b = rgba.b.toRadixString(16).padLeft(2, '0');
  return '0x${(a + r + g + b).toUpperCase()}';
}

OklchParsed? parseOklch(String str) {
  final regex = RegExp(
      r'oklch\(\s*([\d.]+)%?\s+([\d.]+)\s+([\d.]+)(?:\s*\/\s*([\d.]+))?\s*\)',
      caseSensitive: false);
  final match = regex.firstMatch(str);
  if (match == null) return null;

  final lStr = match.group(1)!;
  final isPercent = str.substring(match.start, match.end).contains('%');
  double l = double.parse(lStr);
  if (isPercent) l /= 100.0;

  final c = double.parse(match.group(2)!);
  final h = double.parse(match.group(3)!);
  final opacityStr = match.group(4);
  final opacity = opacityStr != null ? double.parse(opacityStr) : 1.0;

  final rgb = oklchToRgb(l, c, h, opacity);

  return OklchParsed(
    r: rgb.r,
    g: rgb.g,
    b: rgb.b,
    a: opacity,
    l: l,
    c: c,
    h: h,
    opacity: opacity,
  );
}
